tests/unit/plugins/modules/FakeAnsibleModule.py

```diff
@@ -22,10 +26,55 @@ class FailJsonException(AnsibleModuleException):
     pass
 
 
+class CommandNotRegisteredException(Exception):
+    pass
+
+
 class FakeAnsibleModule:
     def __init__(self, params=None, check_mode=False):
         self.params = params
         self.check_mode = check_mode
+        self.commands = defaultdict(list)
+        self._keep_last_command = False
+
+    def register_command(self, args, rc=0, stdout='', stderr=''):
+        """Register a command which will be 'executed' later with run_command.
+
+        Each command you register is only used once, so you can register the
+        same command multiple times with different results. If you call
+        `keep_last_command(True)` than the last instance of any particular
+        command will be preserved and used as the result of any subsequent
+        invocations."""
+        if not isinstance(rc, int):
+            raise TypeError('rc must be int, not "%s"' % (rc,))
+        self.commands[tuple(args)].append((rc, stdout, stderr))
+
+    def keep_last_command(self, value):
+        """Specify whether to use the last instance of a command for multiple invocations."""
+        self._keep_last_command = value
+
+    # Note: Most of these arguments don't do anything yet in this stub; they're
+    # they're just here for call signature compatibility with the real
+    # function. If you need any of this functionality in your tests feel free
+    # to add it!
+    def run_command(self, args, check_rc=False, close_fds=True,
+                    executable=None, data=None, binary_data=False,
+                    path_prefix=None, cwd=None, use_unsafe_shell=False,
+                    prompt_regex=None, environ_update=None, umask=None,
+                    encoding='utf-8', errors='surrogate_or_strict',
+                    expand_user_and_vars=True, pass_fds=None,
+                    before_communicate_callback=None, ignore_invalid_cwd=True,
+                    handle_exceptions=True):
```

Comment:
Hmm, I'm not really happy about the extensions to `FakeAnsibleModule`. If there would be a derived version from `FakeAnsibleModule` that provides these functionality I think it would be OK, but modifying `FakeAnsibleModule` itself is not a good idea IMO.

I would add a `FakeCommandAnsibleModule` class derived from `AnsibleModule`, that provides this functionality.

Reply:
Can you elaborate on what your concern is here? This change is completely backward compatible and doesn't interfere with any of the previous functionality in `FakeAnsibleModule`, so I'm not sure what there is to gain with creating a separate class for this new functionality. That just seems like clutter.
It's not like anybody could have been using `run_command` in tests that use `FakeAnsibleModule` before I added this change... the method didn't exist since `FakeAnsibleModule` isn't subclassed from anything.

Reply:
AnsibleModule provides a set of methods that might get extended at any time. If they add `register_command` or `keep_last_command`, your extensions suddenly have potential to break other tests / make them more complicated.

Also most modules using `FakeAnsibleModule` use mocks to mock `run_command` if they need it, so adding some generic functions to `FakeAnsibleModule` that only one single module needs and most others don't seems potentially problematic to me.

Reply:
Counter-arguments:

> AnsibleModule provides a set of methods that might get extended at any time. If they add `register_command` or `keep_last_command`, your extensions suddenly have potential to break other tests / make them more complicated.

1. Surely you must admit that this is a stretch.
2. If this happens then my tests will break the build and it's not going to be particularly more work to patch multiple tests that break than it will be to patch one test that breaks.
3. This will be true even if I put the functionality in a separate class.
 
> Also most modules using `FakeAnsibleModule` use mocks to mock `run_command` if they need it, so adding some generic functions to `FakeAnsibleModule` that only one single module needs and most others don't seems potentially problematic to me.

1. I'm sure that the other tests would have been able to benefit from my functionality if it had been present when they were written and could benefit from being updated to use this functionality once it's added.
2. If they're already mocking out `run_command` then my change will have no impact on them because they're replacing the function I added with theirs (and indeed we see that, since all the tests are passing).

Reply:
> I'm sure that the other tests would have been able to benefit from my functionality if it had been present when they were written and could benefit from being updated to use this functionality once it's added.

They can, but they should explicitly opt-in to use it IMO.

Reply:
:shrug: It is _impossible_ that any current test is calling `run_command` on an instance of the `FakeAnsibleModule` class right now, because the method doesn't exist in the current version of the class (prior to my change).

It is therefore _impossible_ that my change will break any current test. Which, again, we've already proven from the fact that the tests all pass with my change on the branch.

Anyone who chooses to call `run_command` on the `FakeAnsibleModule` object with my change will in fact be explicitly opting in to using the functionality I added, because their tests will fail unless they call `register_command` before `run_command`.

In short, I've designed my change in such a way that it is _impossible_ for anyone to use them by accident.

:shrug: I disagree with you here, and I've comprehensively explained why, but ultimately it's your call. If you tell me you won't merge the patch unless I move the functionality into a different class, fine, I'll do that, but I don't think it's the right thing to do.

Reply:
ping

Reply:
My comments was not for existing tests (in this repository), but for future tests and future modifications of ansible-core.

In any case, the code you are modifying here belongs to the xenserver modules (see `git log --stat --follow tests/unit/plugins/modules/FakeAnsibleModule.py`), so adding some specific code to it that's used for your module and might not even be useful for other modules that use `AnsibleModule.run_command` is IMO a bad idea. Adding it with a mix-in class so tests can opt-in to use it is better; or move it to its own class and use `amodule.run_module = InstanceOfThatClass()` to use it, and change `amodule.register_command()` to `amodule.run_command.register_command()`.